[RequirementMachine] Diagnose type parameters that are made concrete by a same-type requirement.

hborla opened the pull request:
This change adds diagnostics for when type parameters in a generic signature are made concrete or equivalent to another type parameter through a same-type constraint by looking for non-canonical generic parameters in a minimal signature.

hborla committed:
[RequirementMachine] Diagnose type parameters that are made concrete by a
same-type requirement.

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test

slavapestov commented on lib/AST/RequirementMachine/RequirementMachineRequests.cpp:
```diff
@@ -850,9 +856,33 @@ InferredGenericSignatureRequestRQM::evaluate(
-    }
 
-    // FIXME: Handle allowConcreteGenericParams
+      if (!allowConcreteGenericParams) {
```
The FIXME was in the wrong place :) This should happen after the installRequirementMachine() call below, otherwise we'll always create a new requirement machine for this signature even though we have one on-hand here.

Alternatively, use the RequirementMachine::getCanonicalTypeInContext() method instead of calling it on the signature.

slavapestov commented on test/Compatibility/accessibility.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-typecheck-verify-swift -swift-version 4 -requirement-machine-protocol-signatures=verify -requirement-machine-inferred-signatures=verify
+// RUN: %target-typecheck-verify-swift -swift-version 4
 // RUN: %target-typecheck-verify-swift -swift-version 4.2 -requirement-machine-protocol-signatures=verify -requirement-machine-inferred-signatures=verify
```
Does the -swift-version 4.2 also pass without verify?

hborla replied:
Yep! Fixed.

slavapestov commented on test/Generics/concrete_contraction_unrelated_typealias.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-swift-frontend -typecheck %s -debug-generic-signatures -requirement-machine-inferred-signatures=on 2>&1 | %FileCheck %s
+// RUN: not %target-swift-frontend -typecheck %s -debug-generic-signatures -requirement-machine-inferred-signatures=on 2>&1 | %FileCheck %s
```
Maybe this test case should use -verify

slavapestov commented on lib/AST/RequirementMachine/RequirementMachineRequests.cpp:
```diff
@@ -413,6 +413,12 @@ RequirementSignatureRequestRQM::evaluate(Evaluator &evaluator,
       diagnoseRequirementErrors(ctx, errors,
                                 AllowConcreteTypePolicy::NestedAssocTypes);
+
+      if (machine->getConcreteType(proto->getSelfInterfaceType(),
```
Do you need to do this for each protocol in a connected component? (I know, it's a ridiculous example that nobody will hit) Maybe try something like this

```swift
protocol P1 where Self == Int {
  associatedtype T : P2
}

protocol P2 where Self == Int {
  associatedtype T : P1
}
```

The requirement signature of P1 and P2 is computed simultaneously.

I think the source location used for conflict diagnostics might be suspect too; we seem to always use proto->getLoc() as well.

hborla replied:
> I think the source location used for conflict diagnostics might be suspect too; we seem to always use proto->getLoc() as well.

I know, right now conflict diagnostics always use the location of the signature itself. Instead, it should try to use the location of the conflicting requirement inside the signature using the structural requirements in the rewrite system.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla committed:
[RequirementMachine] Remove unnecessary conditions in concrete type parameter
diagnostics code.

hborla committed:
[RequirementMachine] Downgrade concrete type parameter diagnostics to a
warning.

These diagnostics are stricter in the RequirementMachine than in the GSB,
and there's code that relies on the more relaxed diagnostics in the source
compatibility suite. Downgrade these diagnostics to a warning using
warnUntilSwiftVersion(6).

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
Source compatibility failed with a UPASS for the SRP project.

@swift-ci please smoke test Linux

hborla commented:
@swift-ci please smoke test Linux
